whistl/classifier.py

```diff
@@ -350,6 +354,181 @@ def train_with_erm(classifier, map_file, train_dirs, tune_dirs, gene_file, num_e
         return results
 
 
+def train_multitask(representation, labels, map_file, data_dirs, gene_file, num_epochs,
+                    loss_scaling_factor, device, save_file_dir=None, logger=None,
+                    tune_study_count=2):
+    '''
+    Train a multitask learning model to classify multiple diseases from gene expression data
+
+    Arguments
+    ---------
+    representation: pytorch.nn.Module to be used as a gene expression representation, for example
+        model.ExpressionRepresentation
+    labels: list of strs
+        The disease labels to train act as different tasks for the model. Note that the label
+        'healthy' should not be in this list, as most studies will contain healthy samples
+    map_file: string or Path
+        The file created by label_samples.py to be used to match samples to labels
+    data_dirs: list of strs or list of Paths
+        The directories containing gene expression data for various experiments
+    gene_file: string or Path
+        The path to the file containing the list of genes to use in the model
+    num_epochs: int
+        The number of times the model should be trained on all the data
+    loss_scaling_factor: float
+        A hyperparameter that balances the classification loss penalty with the study invariance
+        penalty. A larger value of loss_scaling_factor will cause the loss to count for more and
+        the invariance penalty to count for less
+    device: torch.device
+        The device to train the model on (either a gpu or a cpu)
+    save_file_dir: str
+        The directory to save the models to
+    logger: logging.logger
+        The python logger object to handle printing logs
+    tune_study_count: int
+        The number of studies to assign to the tune set
+
+    Returns
+    -------
+    results: dict
+        A dictionary containing lists tracking different loss metrics across epochs
+    '''
+    sample_to_label = util.parse_map_file(map_file)
+
+    num_genes = util.get_gene_count(gene_file)
+    # Initialize the representation portion of the model
+    representation = representation(num_genes).double().to(device)
+
+    task_to_results = {}
+    for task in labels:
```

Comment:
Will the order of the tasks matter here? My understanding of multitask learning was that all of the tasks are trained jointly, so for each training epoch you'd calculate the loss for each task, then add them to get the total loss - are you taking a different approach and I'm just thinking of a particular sub-category of multitask learning?

Reply:
It's possible to train tasks jointly or separately. I did it separately first because I'm more familiar with joint training and wanted to make sure that swapping heads was working correctly